appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/EJBContainerTransactionManager.java

```diff
@@ -428,46 +419,42 @@ protected void useClientTx(Transaction prevTx, EjbInvocation inv) {
      * Handle transaction requirements, if any, after invoking bean method
      */
     protected void postInvokeTx(EjbInvocation inv) throws Exception {
-
-        Throwable exception = inv.exception;
+        LOG.finest(() -> String.format("postInvokeTx(inv=%s)", inv));
 
         // For StatelessSessionBeans, ejbCreate/ejbRemove was called without a Tx,
         // so resume client's Tx if needed.
         // For StatefulSessionBeans ejbCreate/ejbRemove was called with or without a Tx,
         // so resume client's Tx if needed.
         // For EntityBeans, ejbCreate/ejbRemove/ejbFind must be called with a Tx
         // so no special processing needed.
-        if ( container.resumeTransaction(inv) ) {
+        if (container.resumeTransaction(inv)) {
            // check if there was a suspended client Tx
-            if ( inv.clientTx != null ) {
+            if (inv.clientTx != null) {
                 transactionManager.resume(inv.clientTx);
             }
-
-            if ( inv.exception != null
-                     && inv.exception instanceof BaseContainer.PreInvokeException ) {
-                inv.exception = ((BaseContainer.PreInvokeException)exception).exception;
+            if (inv.exception != null && inv.exception instanceof BaseContainer.PreInvokeException) {
+                inv.exception = inv.exception.getCause();
             }
-
             return;
         }
 
         EJBContextImpl context = (EJBContextImpl)inv.context;
 
-        int status = transactionManager.getStatus();
-        int txAttr = inv.invocationInfo.txAttr;
-
-        Throwable newException = exception; // default
+        final int status = transactionManager.getStatus();
+        final int txAttr = inv.invocationInfo.txAttr;
+        final Throwable oriException = inv.exception;
 
         // Note: inv.exception may have been thrown by the container
         // during preInvoke (i.e. bean may never have been invoked).
 
         // Exception and Tx handling rules. See EJB2.0 Sections 17.6, 18.3.
+        final Throwable newException;
```

Comment:
Note: it is assigned only one. It is little bit more code, but it is still more comprehensible than it was before. Only one assignment.

Reply:
I finally moved these checks into the methods in else branches